### DISCUSSION TITLE
Foreign Key Array

### DISCUSSION THREAD

**Discussion by kierangilliam** [supabase/supabase#535] [Questions] [answered]
Is it possible to create a foreign key array?

Using the following setup, I get an error:

![image](https://user-images.githubusercontent.com/18583882/104772470-574e7500-5728-11eb-91fd-a99639f4a282.png)

![image](https://user-images.githubusercontent.com/18583882/104772502-633a3700-5728-11eb-8d27-9d5e6d0c2ff5.png)

(In console)
```
code: 400
message: "foreign key constraint "block_foreign_array_fkey" cannot be implemented"
```

Is this possible to do with Supabase?

**Comment by steve-chavez** [comment, marked as answer]
Hi Kieran,

> Is this possible to do with Supabase?

Not possible, but might be in the future. Foreign Key Arrays is a [work in progress](https://commitfest.postgresql.org/17/1252/) in PostgreSQL.

Alternatively, you can do the usual approach of creating a junction(or join) table for this. Check this [SO question](https://stackoverflow.com/questions/9789736/how-to-implement-a-many-to-many-relationship-in-postgresql).

Doing a many-to-many [embedding](https://supabase.io/docs/client/limit#with-embedded-resources) then will work transparently with supabase-js.

**Reply by carllippert** [reply]
How can I filter by a nested relationship then? 

I moved to text array because I was not able to filter my table by its related table as mentioned in docs. 
Is two .eq not allowed?

ie. 

table1 locations { location_id, public, etc }
table2 tags { tag_id, name, etc }
table 3 location_tags { location_id, tag_id, etc }

query locations ->

.select("*, location_tags ( *,  tag_id ( tag_id,  * ) ) )"
.eq("public", true)
 .eq('location_tags.tag_id.tag_id', <filter_key>)

I would get all the locations no matter what relational results where found inside the location_tags table. 

Any help appreciated!

**Reply by soedirgo** [reply]
@carllippert can you try the new [`!inner`](https://github.com/supabase/postgrest-js/issues/197) filter? Something like:
```js
const { data, error } = await supabase
    .from('locations')
    .select('*, tags!inner(*)')
    .eq('public', true)
    .eq('tags.tag_id', ...)
```

**Reply by carllippert** [reply]
at first I thought it wasn't working. But then I figured it out! thnx @soedirgo!

```js
   let query = supabase
      .from("locations")
      .select("*, location_tags!inner (*, tag_id!inner ( * ))")
      .eq("public", true)
      .limit(5);

    if (filters.length > 0) {
      filters.forEach((filter) => {
        query.eq('location_tags.tag_id.tag_id', filter.tag_id);
      }); 
    }

    const { data, error } = await query;
```

One thing/bug? I did notice is that you can't use typescript the same way I have previously when doing nested filters. 
It gets angry with the query.eq(arg1, arg2) arg1 not being recognized column name. 

i.e. 
```js
 let query = supabase
     .from<definitions['locations']>("locations")
```

**Reply by soedirgo** [reply]
Ah, yeah... that wasn't accounted for when the types were first designed. Filing this on https://github.com/supabase/supabase-js/issues/170.
